### ISSUE TITLE
min/max of empty arrays 

### ISSUE THREAD

**Issue by bsipocz** [numpy/numpy#5032]
A similar issue was raised in #2670, however the edge case of (0, ) shape is still not covered. I can go around it with checking for non-zero length in an `if` statement, but it would be more practical if `np.min` could handle this, too.

```
In [24]: [np.min(param[x][1]) for x in fields]
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-24-076aa74e3684> in <module>()
----> 1 [np.min(param[x][1]) for x in fields]

/data/wts/pipeline64/python/lib/python2.6/site-packages/numpy/core/fromnumeric.pyc in amin(a, axis, out)
   1893     except AttributeError:
   1894         return _wrapit(a, 'min', axis, out)
-> 1895     return amin(axis, out)
   1896 
   1897 

ValueError: zero-size array to minimum.reduce without identity

In [25]: param[x]
Out[25]: 
{1: array([], dtype=float64),
 2: array([], dtype=float64),
 3: array([], dtype=float64),
 4: array([], dtype=float64)}

In [26]: param[x][1].shape
Out[26]: (0,)
```

**Comment by RerumNovarum** [comment]
I believe it'd be perfectly fine to provide an option to manually specify meaningful identity in a min/max/reduce call.
Like `np.zeros((0,0)).min(identity=np.inf)` yields infty,
while `np.zeros((0,0)).min()` can still throw a `ValueError` or whatever.

**Comment by eric-wieser** [comment]
Before anything like that can be implemented, #8955 needs to be merged

**Comment by hameerabbasi** [comment]
Repeating from #8955:

I would suggest adding identities to np.minimum and np.maximum, which would be +inf and -inf respectively. This is intuitively and logically true, as max(a, -inf) = a and min(a, +inf) = a.

I agree this probably doesn't match Python conventions but does make sense mathematically. Not to mention, np.nanmin(all-nan) and np.min([]) should be `np.inf` (Mathematically speaking).

I agree it depends on how you define it, but this is probably the most widely accepted view.

**Comment by hameerabbasi** [comment]
> min(np.array(data, int16)).dtype must return an int16 even when len(data) == 0 - so returning inf is not possible

I guess you could cast and then return whatever is the maximum of `int16`, which would be valid too (as the universal set is the set of all `int16`). I guess it comes down to whether we want the maximum or the supremum.

> max(abs(complex_vals)) has a natural identity of 0, not -infinity, but there's no way for max to distinguish that case. Elsewhere a max_abs ufunc has been proposed to solve that problem.

I would still argue that the identity value should be `-inf` in this case, and we should aim for `np.max(np.abs(complex_vals), identity=0)`.

**Comment by eric-wieser** [comment]
> I would still argue that the identity value should be -inf in this case

I think this would be too surprising to be useful. Expressed formally, we can cover all the cases:

* Let X be some set of values
* Let Y be N samples from X
* Expectation: predicate(min(Y)) is true

The only way this can hold when N == 0 is if min([]) == min(X), but `min` has no way of know what `min(X)` is. When X is ℝ, then the answer is `-inf`. But in general, numpy shouldn't assume the invariants of the set of values the user is passing.

To quote the zen of python

> In the face of ambiguity, refuse the temptation to guess.

So I'm also -1 on a default for empty arrays.

I'm +1 on allowing `np.max(np.abs(complex_vals), identity=0)`. I though that was blocked by gh-8955, but actually I think it was only blocked by gh-8952 (now merged), so a patch would be welcome!

**Comment by hameerabbasi** [comment]
I'll look into it. I'm not too deep into the annals of core Numpy, if you could point me towards relevant files that'd be a big help. As long as they're Cython/Pure C++ I should be able to do it, but if they're Python API C++... I'm not entirely sure.

**Comment by eric-wieser** [comment]
The function I mention is defined here

https://github.com/numpy/numpy/blob/e6956dde2299fb9980eb00437a1cd3af97da5958/numpy/core/src/umath/ufunc_object.c#L2833-L2898

And called here:

https://github.com/numpy/numpy/blob/e6956dde2299fb9980eb00437a1cd3af97da5958/numpy/core/src/umath/ufunc_object.c#L3632-L3914

You'd need to:
* Parse a `identity` kwarg for `reduce` in one of the `PyArg_ParseTupleAndKeywords` calls, and call `_get_identity` if the argument is missing.
* Pass it to `PyUFunc_Reduce`, rather than having PyUFunc_Reduce compute it internally
* Add tests and documentation

**Comment by hameerabbasi** [comment]
Why leave it at `reduce` when `reduceat`, `accumulate` and `accumulateat`, and NaN-aggregations would also need it? This is rather big, but I'm willing to take it on. 😄

**Comment by eric-wieser** [comment]
* `reduceat` doesn't even use it's identity - don't open that can of worms. There's another issue about that elsewhere
* `accumulate` perhaps doesn't need it? `np.minimum.accumulate([])` is already well-defined
* `accumulateat` doesn't exist

**Comment by hameerabbasi** [comment]
Can you reopen the issue?

**Comment by carlosgmartin** [comment]
FWIW, I agree with @hameerabbasi. Just as the
* sum of an empty array is the [identity element](https://ncatlab.org/nlab/show/identity+element) of addition, which is [zero](https://ncatlab.org/nlab/show/zero)
* prod of an empty array is the identity element of multiplication, which is [one](https://ncatlab.org/nlab/show/one)

It ought to be the case that the
* max of an empty array is the identity element of [max](https://ncatlab.org/nlab/show/join), which is the [least element](https://ncatlab.org/nlab/show/bottom) of its dtype
* min of an empty array is the identity element of [min](https://ncatlab.org/nlab/show/meet), which is [greatest element](https://ncatlab.org/nlab/show/top) of its dtype

This is the mathematically correct way to treat the [supremum](https://ncatlab.org/nlab/show/join)/[infimum](https://ncatlab.org/nlab/show/meet) of an empty collection.

Related: https://github.com/google/jax/issues/18661.

**Comment by jakevdp** [comment]
Echoing some of my thoughts from #18661 here in case it is helpful:

The problem with using the identity for min/max is that the supremum and infimum are only well-defined for bounded sets. Real numbers and integers are unbounded sets, so the supremum and infimum are not defined. Well it's true that fixed-width approximations to these sets (like the sets of values representable by `float64` or `int64`) are bounded, it would be confusing to use their supremum or infimum as identities of min and max, because these are properties of the implementation, not properties of the numbers that the implementation represents (and no, $\infty$ is not a real number 😁).

Contrast this with `sum`/`prod`, in which the identities *are* well-defined for integers and real numbers, so it makes sense to return those identities in the case of a reduction over an empty array.

**Comment by rkern** [comment]
FWIW, IEEE 754 floating point numbers model the [affinely extended reals](https://mathworld.wolfram.com/AffinelyExtendedRealNumbers.html) more so than the reals, per se, and -∞ and +∞ are perfectly fine members and indeed the well-defined infimum and supremum identity values for `min` and `max`.

While I don't think it's _necessary_ to define the result of `min` and `max` on empty typed arrays like that, I think it'd be valid and no more confusing than any of the other places where we use the properties of the specific `dtype`. For the integer dtypes, the finite bounds come into play elsewhere; I'm not sure why it would be confusing to use them in a mathematical place where such bounds are quite salient.

**Comment by jakevdp** [comment]
OK, fair enough. But I do think it would be quite confusing to the majority of NumPy users if `np.arange(0).min()` were to return `9223372036854775807`

**Comment by SomeoneSerge** [comment]
+1 to throwing an exception but accepting an optional `identity=`

**Comment by jakevdp** [comment]
> +1 to throwing an exception but accepting an optional `identity=`

Note that the identity argument is already available as of NumPy 1.15 (but it's called `initial`):
```python
>>> np.arange(0).min(initial=np.iinfo('int64').max)
9223372036854775807
```

**Comment by seberg** [comment]
> A possible compromise / feature that occurs to me is adding .max and .min attributes

I have made a lot of work to allow adding such attributes to the dtype instances (not sure I like `min/max`, it feels like `.info.min` or just something longer may be better).
In fact, I always thought that is where finfo/iinfo must move long term.  So I would welcome work for going towards that.

> but it's called initial

Right, just to note that the name is of course intentional, because `identity` hints to unrelated properties.  Also I want(ed) the freedom to add a `default=` which is *not* used as the initial value for the reduce-like operation but returned for empty results.  That would allow `default=np.nan` to return `NaN` for empty min.

---

Besides boolean, where I guess it is fully unambigious.  It seems to me that we will remain at: yes, it is plausible to use `±inf`, but it is unclear that it is better.

I do think both are valid choices, but it still seems like more guesswork to me than others (like `mean([])` returning `NaN`, because it does `sum([])/0 = 0/0`).

Probably the reason is that within many contexts the actual minimum value may be well defined, e.g. `0`.  Also, I would think you should give an "infinity created" (division by zero) warning, which would be tedious ;).

**Comment by carlosgmartin** [comment]
> > A possible compromise / feature that occurs to me is adding .max and .min attributes
> 
> I have made a lot of work to allow adding such attributes to the dtype instances (not sure I like `min/max`, it feels like `.info.min` or just something longer may be better). In fact, I always thought that is where finfo/iinfo must move long term. So I would welcome work for going towards that.

I'm interested in a uniform interface for accessing the minimum/maximum value of a given dtype ([JAX example](https://github.com/google/jax/issues/18661#issuecomment-1829031914)). Is there an open issue discussing this? If not, I could open one.

**Comment by seberg** [comment]
Not sure, https://github.com/numpy/numpy/issues/27231 is somewhat related, though.  You can open a new one and even work on it if you like.

(I am not sure how simple it would be to add these attributes.  It might be very weedy in C, but actually, I think it may just be possible to do it in `np.dtypes.` in Python.  In which case the only complicated thing is to make the decision.)
